KAFKA-12234: Implement request/response for offsetFetch batching (KIP-709)

skaundinya15 opened the pull request:
This implements the request and response portion of KIP-709: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=173084258. The Admin APIs will be implemented in https://github.com/apache/kafka/pull/10964.

It updates the OffsetFetch request and response to support fetching offsets for multiple consumer groups at a time. If the broker does not support the new OffsetFetch version, clients can revert to the previous behaviour and use a request for each coordinator.

We are trying to get this change into the 3.0 release. We believe this change will be safe to check in as we are changing the request protocol, and not the underlying Admin APIs. As a result, we are not changing any major client side code, just updating the protocol for the brokers to support multiple groups for a `offsetFetch` request. We have also added substantial request/response testing to ensure we are having backward compatibility.

skaundinya15 committed:
KIP-709: Implement request/response for offsetFetch batching

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListConsumerGroupOffsetsHandler.java:
```diff
@@ -73,7 +73,10 @@ public String apiName() {
     public OffsetFetchRequest.Builder buildRequest(int coordinatorId, Set<CoordinatorKey> keys) {
         // Set the flag to false as for admin client request,
         // we don't need to wait for any pending offset state to clear.
-        return new OffsetFetchRequest.Builder(groupId.idValue, false, partitions, false);
```
Given that single groupid is a common pattern, we could just retain the old constructor as well.

skaundinya15 replied:
The old `Builder` still exists, I just changed it to use the new format, since when we implement the Admin APIs we will have to change everything to use that anyway. However if you think it may be better to retain the old pattern I can change it - what do you think?

rajinisivaram replied:
We can leave the admin client changes for the next PR to keep the changes in this one small for 3.0.

skaundinya15 replied:
Okay, I have removed it.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListConsumerGroupOffsetsHandler.java:
```diff
@@ -87,12 +90,26 @@ public String apiName() {
-        if (response.error() != Errors.NONE) {
-            handleError(groupId, response.error(), failed, unmapped);
+        Errors responseError = response.error();
+        // check if error is null, if it is we are dealing with v8 response
```
Can we move this logic to the response class?

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListConsumerGroupOffsetsHandler.java:
```diff
@@ -87,12 +90,26 @@ public String apiName() {
+            Map<TopicPartition, OffsetFetchResponse.PartitionData> responseData =
+                response.responseData(groupId.idValue) == null ? response.oldResponseData() :
+                    response.responseData(groupId.idValue);
+            for (Map.Entry<TopicPartition, OffsetFetchResponse.PartitionData> entry : responseData.entrySet()) {
```
As before, it will be good if we can move the version logic to the response.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1308,29 +1308,41 @@ private OffsetFetchResponseHandler() {
-                log.debug("Offset fetch failed: {}", error.message());
+            Errors responseError = response.error();
+            // check if error is null, if it is we are dealing with v8 response
+            if (responseError == null) {
```
We can move this to the response object, otherwise it is duplicated from above.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1308,29 +1308,41 @@ private OffsetFetchResponseHandler() {
+            // if map entry is null, we know we are handling a response less than V8
+            boolean useV8 = response.responseData(rebalanceConfig.groupId) != null;
+            Map<TopicPartition, OffsetFetchResponse.PartitionData> responseData = useV8 ?
+                response.responseData(rebalanceConfig.groupId) : response.oldResponseData();
```
As before, it will be good if we don't have repeated logic like this in every client.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -46,9 +52,9 @@
-                       boolean throwOnFetchStableOffsetsUnsupported) {
+            boolean requireStable,
+            List<TopicPartition> partitions,
+            boolean throwOnFetchStableOffsetsUnsupported) {
```
nit: unnecessary indentation change

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -68,36 +74,117 @@ public Builder(String groupId,
-                            .setTopics(topics);
+                .setGroupId(groupId)
+                .setRequireStable(requireStable)
+                .setTopics(topics);
```
nit: unnecessary indentation change

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -68,36 +74,117 @@ public Builder(String groupId,
+            List<OffsetFetchRequestGroup> groups = new ArrayList<>();
+            for (Entry<String, List<TopicPartition>> entry : groupIdToTopicPartitionMap.entrySet()) {
+                final List<OffsetFetchRequestTopics> topics;
+                if (groupIdToTopicPartitionMap.get(entry.getKey()) != null) {
```
Use `entry.getValue()`?

skaundinya15 replied:
Here I'm using `entry.getKey()` because we are trying to get the list of topic partitions for a specific group id. `entry.getValue()` would give us `List<TopicPartition>` whereas we need `String`, which `entry.getKey()` would give us.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -68,36 +74,117 @@ public Builder(String groupId,
+                if (groupIdToTopicPartitionMap.get(entry.getKey()) != null) {
+                    Map<String, OffsetFetchRequestTopics> offsetFetchRequestTopicMap =
+                        new HashMap<>();
+                    for (TopicPartition topicPartition : groupIdToTopicPartitionMap.get(entry.getKey())) {
```
As before, `entry.getValue()`, we should just store that in a local variable

skaundinya15 replied:
Good point on storing it in the local variable, I will do that. But I think we still do want `entry.getKey()` in this situation, not `entry.getValue()`.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -68,36 +74,117 @@ public Builder(String groupId,
+                        .setTopics(oldFormatTopics)
+                        .setRequireStable(data.requireStable());
+                }
+                if (isAllTopicPartitions() && version < 2) {
```
This could be right at the top as it was before to avoid this big `if block`

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -68,36 +74,117 @@ public Builder(String groupId,
+                    throw new UnsupportedVersionException("The broker only supports OffsetFetchRequest " +
+                        "v" + version + ", but we need v2 or newer to request all topic partitions.");
+                }
+                if (data.requireStable() && version < 7) {
```
Again better to move this to the top as well so that we have smaller sequence of if statements rather than nested ifs.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -128,6 +227,38 @@ public boolean requireStable() {
+
+    public Map<String, List<OffsetFetchRequestTopics>> groupIdsToTopics() {
+        Map<String, List<OffsetFetchRequestTopics>> groupIdsToTopics = new HashMap<>();
+        for (OffsetFetchRequestGroup group : data.groupIds()) {
```
Could use stream()?

skaundinya15 replied:
`Collectors.toMap()` doesn't allow `null` entries (we could have `null` entries if a specific group wants all the topic partition offsets) and throws an NPE, so instead I can use `foreach` and do an inline put for all entries to the hashmap.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -152,13 +283,27 @@ public OffsetFetchResponse getErrorResponse(int throttleTimeMs, Errors error) {
+        if (version() >= 3 && version() < 8) {
+            return new OffsetFetchResponse(throttleTimeMs, error, responsePartitions);
+        }
+        List<String> groupIds =
```
We have a method above that does this

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -174,6 +319,10 @@ public boolean isAllPartitions() {
         return data.topics() == ALL_TOPIC_PARTITIONS;
     }
 
+    public List<OffsetFetchRequestTopics> isAllPartitionsForGroup() {
```
Name of method suggests we are returning a boolean, but we are returning null. Why do we need a public method that always returns null?

skaundinya15 replied:
Ah good catch - I had originally had it the way where a groupId was passed in and we would return `true` or `false` based on whether the corresponding `List<OffsetFetchRequestTopics>` was `null` or not, but had reverted it when I was trying to debug the test failures. Guess I forgot to put it back to what it was - will change it back.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -65,6 +69,8 @@
     private final OffsetFetchResponseData data;
     private final Errors error;
+    private final Map<String, Errors> groupLevelErrors = new HashMap<>();
+    private final Map<String, Map<TopicPartition, PartitionData>> groupToPartitionData = new HashMap<>();
```
Why are we caching these when we have them in `data`?

skaundinya15 replied:
Yeah, thinking about this more we can probably get rid of `groupToPartitionData`, and build the map every time it gets queried. However I think it is still good to have the `groupLevelErrors` around so it's quick to return the group-level error code. It will follow the similar precedent where previously we cached the overall error code, and instead here we are caching the group level code. What do you think?

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -154,14 +166,88 @@ public OffsetFetchResponse(int throttleTimeMs, Errors error, Map<TopicPartition,
+     * @param errors Error code on a per group level basis
+     * @param responseData Fetched offset information grouped group id
+     */
+    public OffsetFetchResponse(Map<String, Errors> errors, Map<String, Map<TopicPartition, PartitionData>> responseData) {
```
Why do we have this constructor?

skaundinya15 replied:
I added this constructor to keep the precedent of having a constructor with no throttle time and just passing in the `DEFAULT_THROTTLE_TIME` - if it's not needed it can definitely be removed.

rajinisivaram replied:
The other constructor without throttle time is for versions which didn't have throttle time. For newer constructors that will be used with newer versions, we shouldn't need that.

skaundinya15 replied:
Removed the constructor with `DEFAULT_THROTTLE_TIME`.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -154,14 +166,88 @@ public OffsetFetchResponse(int throttleTimeMs, Errors error, Map<TopicPartition,
+    /**
+     * Constructor with throttle time for version 8 and above.
+     * @param throttleTimeMs The time in milliseconds that this response was throttled
+     * @param errors Potential coordinator or group level error code (for api version 2 and later)
```
why api version 2 and later? This constructor is or version 8 and above?

skaundinya15 replied:
Good point, will change it.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -154,14 +166,88 @@ public OffsetFetchResponse(int throttleTimeMs, Errors error, Map<TopicPartition,
+        for (Entry<String, Map<TopicPartition, PartitionData>> entry : responseData.entrySet()) {
+            Map<String, OffsetFetchResponseTopics> offsetFetchResponseTopicsMap = new HashMap<>();
+            for (Entry<TopicPartition, PartitionData> partitionEntry :
+                responseData.get(entry.getKey()).entrySet()) {
```
`entry.getValue()`

skaundinya15 replied:
I think we still want `entry.getKey()` as that returns the `String` group Id. `entry.getValue()` returns `Map<TopicPartition, PartitionData>`. I will change it so that we save `entry.getKey()` in a local variable so we don't have to keep calling it.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -185,21 +271,42 @@ public boolean hasError() {
     }
 
-    public Map<TopicPartition, PartitionData> responseData() {
+    public Map<TopicPartition, PartitionData> oldResponseData() {
```
Is it an odd method name for a public method that we would use elsewhere. What happens when we add other versions - old/older/oldest?

skaundinya15 replied:
Good point, I will rename this method to reflect which versions this applies to and add a comment that it is only public for testing, as I will refactor everything else that uses this to use `responseData()` since we are moving the version checks into that method.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -214,6 +321,10 @@ public Errors error() {
         return responseData;
     }
 
+    public Map<TopicPartition, PartitionData> responseData(String groupId) {
```
We should probably have a `responseData` method that takes version or ensure we have the data cached regardless of version

skaundinya15 replied:
I am planning on changing this method to check and see if `groupToPartitionData` is empty. If it is empty we are using a version less than the one that supports batching and it will fall back to the old behavior. If not we know the response is initialized with data that supports batching and then it returns from the information from `groupToPartitionData`. Do you think this will be sufficient?

rajinisivaram replied:
Yes, sounds good.

rajinisivaram reviewed (commented):
@skaundinya15 Thanks for the PR. I reviewed the implementation and left some comments. Note that the PR build didn't complete because of checkstyle failures. I haven't reviewed all the unit tests yet. As mentioned offline, we need integration tests (tests that extend BaseRequestTest) for batched requests with multiple groups that verify broker's handling of batched requests and as well as authorization.

skaundinya15 commented:
@rajinisivaram Thank you for the reviews, I believe I have addressed all your comments. In the latest commit, I have also fixed some of the failing tests, as well as added additional integration tests for the request + response testing. Please take a look whenever you get a chance - thank you!

skaundinya15 committed:
Addressed review comments, fixed failing tests, added request/response tests

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1308,29 +1308,31 @@ private OffsetFetchResponseHandler() {
-            if (response.hasError()) {
-                Errors error = response.error();
-                log.debug("Offset fetch failed: {}", error.message());
+            Errors responseError = response.groupLevelError(rebalanceConfig.groupId);
```
We could just call this `error` and then won't require the remaining changes below.

skaundinya15 replied:
I had named this `responseError` as there is already an `error` variable defined later in the code to get the partition-level error in the response. Should we still rename it to `error`?

rajinisivaram replied:
we can leave as is.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -78,26 +85,107 @@ boolean isAllTopicPartitions() {
         }
 
+        public Builder(Map<String, List<TopicPartition>> groupIdToTopicPartitionMap,
+            boolean requireStable,
```
nit: indentation

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -78,26 +85,107 @@ boolean isAllTopicPartitions() {
+                }
+                return new OffsetFetchRequest(oldDataFormat == null ? data : oldDataFormat, version);
+            }
+            // version 8 but have used old format of request, convert to version 8 of request
```
May be better to put the following code in an else statement and put the comment at the start of the `if` block since we may be converting for version < 8 as well in the block above.

skaundinya15 replied:
Not sure I totally follow the suggestion here - The `version < 8` block followed by the subsequent `!data.groupIds().isEmpty()` if blog exists to convert the version to an old version if server side we are less than 8. Then we have the additional check for version 8 but using the old version of the request if the client sent the older version but the server supports the newer version, so we convert it to the newer version. What were you specifically thinking to change here to make the code more readable?

rajinisivaram replied:
At the moment, the code seems to do:
```
if (version < 8) {
 do-conversion-if-necessary
}
// version 8 but have used old format of request, convert to version 8 of request
do-conversion-if-necessary
```
My suggestion was to turn it into:
```
// convert data to use the appropriate version since version 8 uses different format
if (version < 8) {
 do-conversion-if-necessary
} else {
 do-conversion-if-necessary
}
```

skaundinya15 replied:
Ah okay, yes that makes sense, will make the change.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -154,14 +167,74 @@ public OffsetFetchResponse(int throttleTimeMs, Errors error, Map<TopicPartition,
+     * @param responseData Fetched offset information grouped by topic-partition and by group
+     */
+    public OffsetFetchResponse(int throttleTimeMs, Map<String, Errors> errors, Map<String,
+        Map<TopicPartition, PartitionData>> responseData) {
```
nit: indentation

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/message/MessageTest.java:
```diff
@@ -662,20 +669,179 @@ public void testOffsetFetchVersions() throws Exception {
-            OffsetFetchResponseData responseData = response.get();
-            if (version <= 1) {
-                responseData.setErrorCode(Errors.NONE.code());
+            if (version < 8) {
```
As before, we should have `if version >= 8` as well.

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/message/MessageTest.java:
```diff
@@ -633,17 +638,19 @@ public void testOffsetFetchVersions() throws Exception {
-                assertThrows(UnsupportedVersionException.class, () -> testAllMessageRoundTripsFromVersion(finalVersion, requireStableData));
-            } else {
-                testAllMessageRoundTripsFromVersion(finalVersion, requireStableData);
+            if (version < 8) {
```
We should probably just add this test for v8 and above as well. The other v8 test can just focus on multiple groups, this one can test the common case of single group that is used by consumers.

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/requests/OffsetFetchRequestTest.java:
```diff
@@ -44,25 +46,20 @@
+    private final String group2 = "group2";
+    private final String group3 = "group3";
+    private final String group4 = "group4";
+    private final String group5 = "group5";
```
Maybe better to define `List<String> groups` to avoid repetition here and in the test logic

skaundinya15 replied:
The reason I had individually defined each group here is because we have to put the groups individually in the map and associated with its specific topic partition. It also made it easier to verify responses after, since we get the partition responses on a group level. Do you have an example of how putting the groups in `List<String>` would help reduce the repetition?

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/requests/OffsetFetchRequestTest.java:
```diff
@@ -76,62 +73,169 @@ public void testConstructor() {
+                assertEquals(Errors.NONE, response.groupLevelError(group2));
+                assertEquals(Errors.NONE, response.groupLevelError(group3));
+                assertEquals(Errors.NONE, response.groupLevelError(group4));
+                assertEquals(Errors.NONE, response.groupLevelError(group5));
```
List of groups will enable us to use loop here.

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/requests/OffsetFetchRequestTest.java:
```diff
@@ -76,62 +73,169 @@ public void testConstructor() {
+                groupPartitionMap.put(group2, null);
+                builder = new Builder(groupPartitionMap, true, false);
+                final short finalVersion = version;
+                assertThrows(NoBatchedOffsetFetchRequestException.class, () -> builder.build(finalVersion));
```
As before, it will be good to include test for `version >=8` as well

skaundinya15 replied:
This specific test checks if the correct exception is thrown if we try to create a batched request for a version less than 8. The `testConstructor` method tests the single group request with version 8 and above. Is there any other case you think we should add for `version >=8`?

rajinisivaram replied:
ah, this is just for batched, no change required.

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/requests/OffsetFetchRequestTest.java:
```diff
@@ -76,62 +73,169 @@ public void testConstructor() {
+    @Test
+    public void testBuildThrowForUnsupportedRequireStable() {
+        for (short version : ApiKeys.OFFSET_FETCH.allVersions()) {
+            if (version < 8) {
```
As before, it will be good to include the same logic for `version >=8` as well. Based on the number of tests that use the pattern of `ApiKeys.OFFSET_FETCH.allVersions()` followed by `if (version < 8)`, it may be good to add the list of versions < 8 and use that in these cases if we require this pattern.

skaundinya15 committed:
Addressed more review comments

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListConsumerGroupOffsetsHandler.java:
```diff
@@ -87,12 +87,16 @@ public String apiName() {
-            for (Map.Entry<TopicPartition, OffsetFetchResponse.PartitionData> entry :
-                response.responseData().entrySet()) {
+            // if entry for group level response data is null, we are getting back an older version
+            // of the response
```
Is this comment required?

skaundinya15 replied:
ah yeah good point, will remove it.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListConsumerGroupOffsetsHandler.java:
```diff
@@ -87,12 +87,16 @@ public String apiName() {
+            // if entry for group level response data is null, we are getting back an older version
+            // of the response
+            Map<TopicPartition, OffsetFetchResponse.PartitionData> responseData =
+                response.responseData(groupId.idValue);
```
Perhaps call the method and variable partitionData or offsetData rather than responseData since responseData is too close to response.data?

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1308,29 +1308,31 @@ private OffsetFetchResponseHandler() {
 
             Set<String> unauthorizedTopics = null;
-            Map<TopicPartition, OffsetAndMetadata> offsets = new HashMap<>(response.responseData().size());
+            Map<TopicPartition, OffsetFetchResponse.PartitionData> responseData =
```
as before, rename responseData?

skaundinya15 replied:
What should we rename this to? I guess technically this would be the `responseData` as in the next line we extract the actual `offsets` map that is a`Map<TopicPartition, OffsetAndMetadata>`.

rajinisivaram replied:
ok, let's leave as is.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -174,6 +315,28 @@ public boolean isAllPartitions() {
+            .groupIds()
+            .stream()
+            .filter(g -> g.groupId().equals(groupId))
+            .collect(toSingleton());
```
This is going to throw IllegalStateException if a group is included multiple times in a request. If that is an invalid request, we should throw InvalidRequestException before it gets here to ensure we don't return UNKNOWN_SERVER_ERROR. More likely, it is valid and we should not throw an exception.

skaundinya15 replied:
Not sure I follow - are you saying that if there is a group that is included multiple times in the request the `stream()` here would throw an `IllegalStateException`? Are you saying we should put a `try...catch` statement here to catch any `IllegalStateException`?

rajinisivaram replied:
The helper method `toSingleton()` throws IllegalStateException if the list size is greater than one. If a request contains the same group twice, it can appear twice in the list.

skaundinya15 replied:
Ah gotcha, makes sense. For now I have removed the `toSingleton()` helper method and collected it into a list that gets the first element instead.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -174,6 +315,28 @@ public boolean isAllPartitions() {
+    }
+
+    // Custom collector to filter a single element
+    private <T> Collector<T, ?, T> toSingleton() {
```
We seem to have two copies of this method in the PR. We need to decide first whether this logic is correct first. If we really do want the method, we need to make it shared.

skaundinya15 replied:
Makes sense, for now I will remove it and convert the methods that used this to collect a `List` and get the first element in the list.

rajinisivaram replied:
Can we a add a test in the new `kafka.server.OffsetFetchRequestTest`? Can do that in a separate PR.

skaundinya15 replied:
Would this be a test for single group requests with v8? I have added that case to the existing single group tests for older versions in `OffsetFetchRequestTest.scala` and will push those changes shortly. Is there another type of test you were thinking of?

rajinisivaram replied:
Same test class `OffsetFetchRequestTest.scala`, a test with v8 with a batched request where the same group appears twice, perhaps with different topics. The response should be either InvalidRequestException because we want to treat it as an error OR actual offsets because we handle the request correctly.

skaundinya15 replied:
Ah okay, good point. Let's do this in a follow up PR, since we are trying to get this feature into 3.0 right now.

skaundinya15 replied:
@rajinisivaram I'm not sure how we could test that case, as the `Builder` for the `OffsetFetchRequest` takes in a `Map<String, List<TopicPartition>>` which means all of the keys have to be unique. As a result if a group appears twice, the latest entry will take precedence. Not sure if this is how we want to handle it, but generally speaking I think we don't support having the same name group id appearing twice in the request - as in it's not possible to construct a request like that, at least not using the `Builder`

rajinisivaram replied:
You can probably set the list in `OffsetFetchRequestData` directly in the test. But let's leave that for a follow-on PR.

skaundinya15 replied:
Opened https://issues.apache.org/jira/browse/KAFKA-13045 to track this.

skaundinya15 replied:
Opened https://github.com/apache/kafka/pull/11000 for this. We can continue the discussion there.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -113,8 +120,14 @@ public int hashCode() {
+
     /**
-     * Constructor for all versions without throttle time.
+     * Constructor without throttle time for version 0 to version 7.
```
It is not actually for version 0 to 7, it is for the versions without throttle time.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -185,21 +249,46 @@ public boolean hasError() {
 
-    public Map<TopicPartition, PartitionData> responseData() {
+    //public for testing purposes
+    public Map<TopicPartition, PartitionData> responseDataV0ToV7() {
```
package-private should be sufficient for unit tests? May also be possible keep this private and call `responseData()` directly to ensure the right data is returned based on version.

skaundinya15 replied:
I'll make it package-private for the unit tests, as the old version of the response did not take in a `groupId` and in order to call the newer function we need to pass in a `groupId` to get the right `Map<TopicPartition, PartitionData>`

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -214,6 +303,46 @@ public Errors error() {
+    }
+
+    // Custom collector to filter a single element
+    private <T> Collector<T, ?, T> toSingleton() {
```
See earlier note about this method.

rajinisivaram commented on clients/src/main/resources/common/message/OffsetFetchResponse.json:
```diff
@@ -30,30 +30,57 @@
+            "about": "The partition-level error code, or 0 if there was no error." }
+        ]}
+      ]},
+      { "name": "ErrorCode", "type": "int16", "versions": "8+", "default": "0", "ignorable": true,
```
Don't think this should be ignorable

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -1580,21 +1623,83 @@ private OffsetCommitResponse createOffsetCommitResponse() {
+                "group1",
+                requireStable,
+                Collections.singletonList(
+                    new TopicPartition("test11", 1)),
```
nit: unnecessary newline

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -1580,21 +1623,83 @@ private OffsetCommitResponse createOffsetCommitResponse() {
+            Collections.singletonMap(
+                "group1",
+                Collections.singletonList(
+                    new TopicPartition("test11", 1))),
```
nit: unnecessary newlines

rajinisivaram commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1255,81 +1256,143 @@ class KafkaApis(val requestChannel: RequestChannel,
     def createResponse(requestThrottleMs: Int): AbstractResponse = {
       val offsetFetchResponse =
-        // reject the request if not authorized to the group
+      // reject the request if not authorized to the group
```
nit: revert indentation change

rajinisivaram commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -1358,17 +1367,241 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+    addAndVerifyAcls(Set(new AccessControlEntry(clientPrincipalString, WildcardHost, READ, ALLOW)), groupFiveResource)
+    addAndVerifyAcls(Set(new AccessControlEntry(clientPrincipalString, WildcardHost, READ, ALLOW)), topicOneResource)
+    addAndVerifyAcls(Set(new AccessControlEntry(clientPrincipalString, WildcardHost, READ, ALLOW)), topicTwoResource)
+    addAndVerifyAcls(Set(new AccessControlEntry(clientPrincipalString, WildcardHost, READ, ALLOW)), topicThreeResource)
```
Looks like they are are all the same with different resources, better to put them in a collection and call `addAndVerifyAcls` over the collection

rajinisivaram commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -1358,17 +1367,241 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+    val offset = 15L
+    val leaderEpoch: Optional[Integer] = Optional.of(1)
+    val metadata = "metadata"
+    val topicOneOffsets = topic1List.asScala.map{
```
nit: space before braces (multiple places)

rajinisivaram commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -1358,17 +1367,241 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+    var consumer = createConsumer()
+    consumer.assign(topic1List)
+    consumer.commitSync(topicOneOffsets)
+    consumer.close()
```
Create a nested method to do this and call it for all cases.

rajinisivaram commented on core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala:
```diff
@@ -0,0 +1,227 @@
+  }
+
+  @Test
+  def testOffsetFetchRequestLessThanV8(): Unit = {
```
Can we make this for all versions with a single group id? That is the scenario in the consumer and we want to make sure the methods in request/response classes handle any version for that case.

skaundinya15 replied:
Makes sense, I will make this method test the single group case.

rajinisivaram commented on core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala:
```diff
@@ -0,0 +1,227 @@
+    val groupTwo = "group2"
+    val groupThree = "group3"
+    val groupFour = "group4"
+    val groupFive = "group5"
```
If we move the single group case to the previous test, we can make this one work on a collection of groups. Instead of named groups like `groupOne`, we can just use group(1).

skaundinya15 replied:
I refactored this test class a bit to make use of more helper methods - is there something additional we can do here to make it work on a collection of groups? Are you suggesting adding an additional method that can create some number of groups that we can use the test with?

rajinisivaram replied:
We can do:
```
val groups = (0 until 5).map(i => s"group$i")
```
And use groups(0), groups(1) etc. instead of groupOne, groupTwo.

rajinisivaram reviewed (commented):
@skaundinya15 Thanks for the updates and tests. Left some more comments. There is a lot of repetitive code, it would be good to minimize those. Left comments for some of them, but it will be good to look at all tests and use collections, helper methods and nested methods in scala rather than repeat statements or group of statements a lot of times for batch tests.

skaundinya15 committed:
Addressed more review comments

skaundinya15 commented:
@kkonstantine This is the PR I am trying to get in time for 3.0, if you will allow for extensions.

skaundinya15 commented:
@rajinisivaram Thank you for the reviews, I have updated the PR addressing all your comments. Whenever you get a chance, the PR is ready for review - thank you!

rajinisivaram commented on clients/src/main/resources/common/message/OffsetFetchRequest.json:
```diff
@@ -31,19 +31,33 @@
+      { "name": "PartitionIndexes", "type": "[]int32", "versions": "0-7",
         "about": "The partition indexes we would like to fetch offsets for." }
     ]},
+    { "name": "GroupIds", "type": "[]OffsetFetchRequestGroup", "versions": "8+",
```
Should we call this `Groups` rather than `GroupIds` since it is not just the group id?

skaundinya15 replied:
Sounds good, will change it.

rajinisivaram commented on clients/src/main/resources/common/message/OffsetFetchResponse.json:
```diff
@@ -30,30 +30,57 @@
-      "about": "The top-level error code, or 0 if there was no error." }
+    { "name": "ErrorCode", "type": "int16", "versions": "2-7", "default": "0", "ignorable": true,
+      "about": "The top-level error code, or 0 if there was no error." },
+    {"name": "GroupIds", "type": "[]OffsetFetchResponseGroup", "versions": "8+",
```
As with the response, should we call this `Groups` rather than `GroupIds`?

skaundinya15 replied:
Sounds good, will change it.

rajinisivaram commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -1358,17 +1367,233 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+    val groupFour = "group4"
+    val groupFourResource = new ResourcePattern(GROUP, groupFour, LITERAL)
+    val groupFive = "group5"
+    val groupFiveResource = new ResourcePattern(GROUP, groupFive, LITERAL)
```
We could do something like:
```
val groups = (0 until 5).map(i => s"group$i")
val groupResources = groups.map(group => new ResourcePattern(GROUP, group, LITERAL))
```

rajinisivaram commented on core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala:
```diff
@@ -0,0 +1,241 @@
+import scala.jdk.CollectionConverters._
+import java.util.{Optional, Properties}
+
+class OffsetFetchRequestTest extends BaseRequestTest{
```
nit: space before {

rajinisivaram reviewed (approved):
@skaundinya15 Thanks for the updates. Left a few minor comments. We can address those in a follow-on PR if you prefer.

skaundinya15 committed:
Addressed more review comments

rajinisivaram commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -1358,17 +1367,222 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+
+    // create group to partition map to build batched offsetFetch request
+    val groupToPartitionMap = new util.HashMap[String, util.List[TopicPartition]]()
+    groupToPartitionMap.put(groups(1), topic1List)
```
groups(0) to group(4)

rajinisivaram commented on core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala:
```diff
@@ -0,0 +1,237 @@
+    // create group to partition map to build batched offsetFetch request
+    val groupToPartitionMap: util.Map[String, util.List[TopicPartition]] =
+      new util.HashMap[String, util.List[TopicPartition]]()
+    groupToPartitionMap.put(groups(1), topic1List)
```
groups(0) to groups(4)

rajinisivaram reviewed (commented):
@skaundinya15 Thanks for the updates. Left comments about indexing of `groups` (starts from 0) in the tests with the change to collection in the last commit. Apart from that LGTM if tests pass.

skaundinya15 commented:
> @skaundinya15 Thanks for the updates. Left comments about indexing of `groups` (starts from 0) in the tests with the change to collection in the last commit. Apart from that LGTM if tests pass.
Just pushed changes for this to fix the failing tests - thanks for pointing it out

kkonstantine reviewed (approved):
Approved for inclusion in 3.0. Thanks for the summary @skaundinya15

skaundinya15 committed:
fix failing tests

rajinisivaram commented:
@skaundinya15 Thanks for the PR. Test failures (KafkaMetadataLogTest.testDeleteSnapshots, RaftClusterTest.testCreateClusterAndCreateAndManyTopicsWithManyPartitions, QueryableStateIntegrationTest.shouldBeAbleToQueryFilterState) not related. Merging to trunk and 3.0.
